{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1726ddd1",
   "metadata": {},
   "source": [
    "# Transformer model for predicting modalities in scRNA-seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a89a670f",
   "metadata": {},
   "source": [
    "**Authors**<br>Vedu Mallela: GiwoTech, vedu.mallela@gmail.com<br>Simon Lee: UC Santa Cruz, siaulee@ucsc.edu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ce5ff58",
   "metadata": {},
   "source": [
    "# Goal of the code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f5c33c9",
   "metadata": {},
   "source": [
    "**TODO: explain algorithm**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efc46e87",
   "metadata": {},
   "source": [
    "# Libraries "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffcf91ff",
   "metadata": {},
   "source": [
    "Import all files and modules for this competition<br>\n",
    "*below will provide documentation of the following libraries*<br>\n",
    "<br>\n",
    "**scanpy** (**s**ingle **c**ell **an**alysis in **Py**thon) - https://scanpy.readthedocs.io/en/stable/ <br>\n",
    "**anndata** (**ann**otated **data**) - https://anndata.readthedocs.io/en/latest/ <br>\n",
    "**matplotlib** - https://matplotlib.org/ <br>\n",
    "**numpy** - https://numpy.org/doc/stable/ <br>\n",
    "**pandas** - https://pandas.pydata.org/ <br>\n",
    "**logging** - https://docs.python.org/3/howto/logging.html <br>\n",
    "**sklearn** - https://scikit-learn.org/stable/ <br>\n",
    "<br>\n",
    "*code begins here*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8a0c65ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import logging\n",
    "\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.metrics import mean_squared_error"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59dd606",
   "metadata": {},
   "source": [
    "# LOAD DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d91dad96",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/tmp/public/multiome/'\n",
    "outpath ='../out/'\n",
    "adata_gex = ad.read_h5ad(path + \"multiome_gex_processed_training.h5ad\")\n",
    "adata_atac = ad.read_h5ad(path + \"multiome_atac_processed_training.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31e4664e",
   "metadata": {},
   "source": [
    "After successfully loading in the data, we can try to begin plotting the batch for the **Assay for Transposase-Accessible Chromatin using sequencing** (ATAC-seq) and **Gene Expression** (GEX) data on the umap interface. These umap projections will be saved to GEX.pdf and ATAC.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "83dbe4b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc.tl.pca(adata_gex)\n",
    "#sc.pl.umap(adata_gex, color=['batch'],save='_GEX', title='GEX umap Display')\n",
    "#sc.tl.pca(adata_atac)\n",
    "#sc.pl.umap(adata_atac, color=['batch'], layer='counts', save='_ATAC', title='ATAC umap Display')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b0acfe1",
   "metadata": {},
   "source": [
    "Next we are going to check out all the indivdual cell types occuring in both the ATAC and GEX data. This way we can see all the types of cells from this dataset provided."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4504456",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc.tl.pca(adata_gex)\n",
    "#sc.pl.umap(adata_gex, color='cell_type',save='_GEX_ct', title='GEX Cell Type umap')\n",
    "#sc.tl.pca(adata_atac)\n",
    "#sc.pl.umap(adata_atac, color='cell_type',save='_ATAC_ct', title='ATAC Cell Type umap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5516e34b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data set\n",
      "GEX data: Total cells=22463, Number features=12160\n",
      "ATAC data: Total cells=22463, Number Features=53020\n"
     ]
    }
   ],
   "source": [
    "# filter out the data\n",
    "# Convert anndata objects to dataframes and filter.  \n",
    "# Genes that show up in < 1% cells are dropped.\n",
    "# Atac seq data that shows up in < 1% cells are dropped \n",
    "\n",
    "gex_df = adata_gex.to_df()\n",
    "atac_df = adata_atac.to_df()\n",
    "\n",
    "mask = gex_df>0\n",
    "total_cells = gex_df.shape[0]\n",
    "maskdf = mask.sum(axis=0)/total_cells*100 <=1\n",
    "\n",
    "gex_feature_drop  = list(maskdf.loc[maskdf==True].index.values)\n",
    "\n",
    "mask = atac_df>0\n",
    "maskdf = mask.sum(axis=0)/total_cells <=0.01\n",
    "\n",
    "atac_feature_drop  = list(maskdf.loc[maskdf==True].index.values)\n",
    "\n",
    "gex_ = gex_df.drop(columns=gex_feature_drop)\n",
    "atac_ = atac_df.drop(columns=atac_feature_drop)\n",
    "\n",
    "print('Filtered data set')\n",
    "print('GEX data: Total cells=' + str(gex_.shape[0]) + ', Number features=' + str(gex_.shape[1]))\n",
    "print('ATAC data: Total cells=' + str(atac_.shape[0]) + ', Number Features=' + str(atac_.shape[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c04d72ab",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from transformers import AutoTokenizer, AutoModelWithLMHead, T5ForConditionalGeneration\n",
    "from pathlib import Path\n",
    "from sklearn.model_selection import train_test_split\n",
    "import torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae4aa867",
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert to numpy array\n",
    "atac_numpy_array = np.array(gex_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0359f8be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the sequence and split it into chunks\n",
    "def read_seq_split(split_dir): \n",
    "    split_dir = (split_dir)\n",
    "    texts = []\n",
    "    labels = []\n",
    "    for label_dir in [\"gex\", \"atac\"]: # for each label\n",
    "        for text_file in (split_dir/label_dir).iterdir():\n",
    "            texts.append(text_file.read_text())\n",
    "\n",
    "    return texts, labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fbc61cc0",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "text input must of type `str` (single example), `List[str]` (batch or single pretokenized example) or `List[List[str]]` (batch of pretokenized examples).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-8ef0690ad04a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0mtokenizer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mT5Tokenizer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_pretrained\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"t5-small\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0mtrain_encodings\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtokenizer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_gex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtruncation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpadding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;31m#val_encodings = tokenizer(val_texts, truncation=True, padding=True)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;31m#test_encodings = tokenizer(test_texts, truncation=True, padding=True)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/transformers/tokenization_utils_base.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, text, text_pair, add_special_tokens, padding, truncation, max_length, stride, is_split_into_words, pad_to_multiple_of, return_tensors, return_token_type_ids, return_attention_mask, return_overflowing_tokens, return_special_tokens_mask, return_offsets_mapping, return_length, verbose, **kwargs)\u001b[0m\n\u001b[1;32m   2375\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2376\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0m_is_valid_text_input\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2377\u001b[0;31m             raise ValueError(\n\u001b[0m\u001b[1;32m   2378\u001b[0m                 \u001b[0;34m\"text input must of type `str` (single example), `List[str]` (batch or single pretokenized example) \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2379\u001b[0m                 \u001b[0;34m\"or `List[List[str]]` (batch of pretokenized examples).\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: text input must of type `str` (single example), `List[str]` (batch or single pretokenized example) or `List[List[str]]` (batch of pretokenized examples)."
     ]
    }
   ],
   "source": [
    "from transformers import AutoTokenizer, AutoModelWithLMHead, T5ForConditionalGeneration, T5Tokenizer\n",
    "from pathlib import Path\n",
    "from sklearn.model_selection import train_test_split\n",
    "import torch\n",
    "#gex_df = adata_gex.to_df()\n",
    "#X = gex_df.drop(['target'],axis=1).values   # independant features\n",
    "#y = gex_df['target'].values                 # dependant variable\n",
    "#train_texts, val_texts, train_labels, val_labels = train_test_split(x, y, train_texts, train_labels, test_size=.2)\n",
    "#train_text, train_labels = read_seq_split(gex_df)\n",
    "#train_text =  \n",
    "\n",
    "tokenizer = T5Tokenizer.from_pretrained(\"t5-small\")\n",
    "train_encodings = tokenizer(adata_gex, truncation=True, padding=True)\n",
    "#val_encodings = tokenizer(val_texts, truncation=True, padding=True)\n",
    "#test_encodings = tokenizer(test_texts, truncation=True, padding=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b6e29c87",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "expected str, bytes or os.PathLike object, not AnnData",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-bce8f24c9c0e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtrain_texts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrain_labels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mread_seq_split\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_gex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;31m#test_texts, test_labels = read_seq_split('figures/test')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-7-20f16ba11525>\u001b[0m in \u001b[0;36mread_seq_split\u001b[0;34m(split_dir)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# read the sequence and split it into chunks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mread_seq_split\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msplit_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0msplit_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msplit_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mtexts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mlabels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/pathlib.py\u001b[0m in \u001b[0;36m__new__\u001b[0;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1039\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcls\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mPath\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1040\u001b[0m             \u001b[0mcls\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mWindowsPath\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'nt'\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mPosixPath\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1041\u001b[0;31m         \u001b[0mself\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_from_parts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minit\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1042\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_flavour\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_supported\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1043\u001b[0m             raise NotImplementedError(\"cannot instantiate %r on your system\"\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/pathlib.py\u001b[0m in \u001b[0;36m_from_parts\u001b[0;34m(cls, args, init)\u001b[0m\n\u001b[1;32m    680\u001b[0m         \u001b[0;31m# right flavour.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    681\u001b[0m         \u001b[0mself\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__new__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 682\u001b[0;31m         \u001b[0mdrv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mroot\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parse_args\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    683\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_drv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdrv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    684\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_root\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mroot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/pathlib.py\u001b[0m in \u001b[0;36m_parse_args\u001b[0;34m(cls, args)\u001b[0m\n\u001b[1;32m    664\u001b[0m                 \u001b[0mparts\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    665\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 666\u001b[0;31m                 \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfspath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    667\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    668\u001b[0m                     \u001b[0;31m# Force-cast str subclasses to str (issue #21127)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: expected str, bytes or os.PathLike object, not AnnData"
     ]
    }
   ],
   "source": [
    "#train_texts, train_labels = read_seq_split(adata_gex) \n",
    "#test_texts, test_labels = read_seq_split('figures/test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d52ed70d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nCount_peaks</th>\n",
       "      <th>atac_fragments</th>\n",
       "      <th>reads_in_peaks_frac</th>\n",
       "      <th>blacklist_fraction</th>\n",
       "      <th>nucleosome_signal</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>pseudotime_order_ATAC</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAGTTGTCACCCTCAC-1-s1d1</th>\n",
       "      <td>4031.0</td>\n",
       "      <td>5400</td>\n",
       "      <td>0.746481</td>\n",
       "      <td>0.003473</td>\n",
       "      <td>0.642468</td>\n",
       "      <td>Naive CD20+ B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTATGGCCATAACGGG-1-s1d1</th>\n",
       "      <td>8636.0</td>\n",
       "      <td>19266</td>\n",
       "      <td>0.448251</td>\n",
       "      <td>0.003126</td>\n",
       "      <td>1.220679</td>\n",
       "      <td>CD14+ Mono</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGCACACAGGTTAAA-1-s1d1</th>\n",
       "      <td>4674.0</td>\n",
       "      <td>6177</td>\n",
       "      <td>0.756678</td>\n",
       "      <td>0.001284</td>\n",
       "      <td>0.692573</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATTTGGTAATGGAA-1-s1d1</th>\n",
       "      <td>2803.0</td>\n",
       "      <td>4019</td>\n",
       "      <td>0.697437</td>\n",
       "      <td>0.000714</td>\n",
       "      <td>0.633838</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACATAGGTGTCCA-1-s1d1</th>\n",
       "      <td>1790.0</td>\n",
       "      <td>2568</td>\n",
       "      <td>0.697040</td>\n",
       "      <td>0.003352</td>\n",
       "      <td>0.727660</td>\n",
       "      <td>CD16+ Mono</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTAAGCAACTAGGG-8-s3d6</th>\n",
       "      <td>7239.0</td>\n",
       "      <td>10580</td>\n",
       "      <td>0.684216</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0.866142</td>\n",
       "      <td>HSC</td>\n",
       "      <td>0.23367</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTCCTTCGGCTAGC-8-s3d6</th>\n",
       "      <td>16056.0</td>\n",
       "      <td>22771</td>\n",
       "      <td>0.705107</td>\n",
       "      <td>0.000810</td>\n",
       "      <td>0.964006</td>\n",
       "      <td>CD4+ T activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTGCGTTGTTGGA-8-s3d6</th>\n",
       "      <td>8388.0</td>\n",
       "      <td>14137</td>\n",
       "      <td>0.593337</td>\n",
       "      <td>0.000358</td>\n",
       "      <td>1.215227</td>\n",
       "      <td>pDC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCCTCCCAGCCAGTT-8-s3d6</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1381</td>\n",
       "      <td>0.724837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTGAACCATCCCGCT-8-s3d6</th>\n",
       "      <td>3183.0</td>\n",
       "      <td>4684</td>\n",
       "      <td>0.679547</td>\n",
       "      <td>0.001257</td>\n",
       "      <td>0.946565</td>\n",
       "      <td>CD4+ T activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22463 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         nCount_peaks  atac_fragments  reads_in_peaks_frac  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1        4031.0            5400             0.746481   \n",
       "CTATGGCCATAACGGG-1-s1d1        8636.0           19266             0.448251   \n",
       "CCGCACACAGGTTAAA-1-s1d1        4674.0            6177             0.756678   \n",
       "TCATTTGGTAATGGAA-1-s1d1        2803.0            4019             0.697437   \n",
       "ACCACATAGGTGTCCA-1-s1d1        1790.0            2568             0.697040   \n",
       "...                               ...             ...                  ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6        7239.0           10580             0.684216   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6       16056.0           22771             0.705107   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6        8388.0           14137             0.593337   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6        1001.0            1381             0.724837   \n",
       "AGTGAACCATCCCGCT-8-s3d6        3183.0            4684             0.679547   \n",
       "\n",
       "                         blacklist_fraction  nucleosome_signal  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1            0.003473           0.642468   \n",
       "CTATGGCCATAACGGG-1-s1d1            0.003126           1.220679   \n",
       "CCGCACACAGGTTAAA-1-s1d1            0.001284           0.692573   \n",
       "TCATTTGGTAATGGAA-1-s1d1            0.000714           0.633838   \n",
       "ACCACATAGGTGTCCA-1-s1d1            0.003352           0.727660   \n",
       "...                                     ...                ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6            0.000553           0.866142   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6            0.000810           0.964006   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6            0.000358           1.215227   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6            0.000000           0.714286   \n",
       "AGTGAACCATCCCGCT-8-s3d6            0.001257           0.946565   \n",
       "\n",
       "                                cell_type  pseudotime_order_ATAC batch  \n",
       "TAGTTGTCACCCTCAC-1-s1d1     Naive CD20+ B                    NaN  s1d1  \n",
       "CTATGGCCATAACGGG-1-s1d1        CD14+ Mono                    NaN  s1d1  \n",
       "CCGCACACAGGTTAAA-1-s1d1            CD8+ T                    NaN  s1d1  \n",
       "TCATTTGGTAATGGAA-1-s1d1            CD8+ T                    NaN  s1d1  \n",
       "ACCACATAGGTGTCCA-1-s1d1        CD16+ Mono                    NaN  s1d1  \n",
       "...                                   ...                    ...   ...  \n",
       "TAGTAAGCAACTAGGG-8-s3d6               HSC                0.23367  s3d6  \n",
       "TGGTCCTTCGGCTAGC-8-s3d6  CD4+ T activated                    NaN  s3d6  \n",
       "CGCTTGCGTTGTTGGA-8-s3d6               pDC                    NaN  s3d6  \n",
       "ACCCTCCCAGCCAGTT-8-s3d6            CD8+ T                    NaN  s3d6  \n",
       "AGTGAACCATCCCGCT-8-s3d6  CD4+ T activated                    NaN  s3d6  \n",
       "\n",
       "[22463 rows x 8 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ccca32",
   "metadata": {},
   "source": [
    "We are now going to print out the number of observations and features of our GEX and ATAC-seq data. \n",
    "\n",
    "Few things to note before we proceed:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6ff7aad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The GEX data has 22463 observations and 13431 features.\n",
      "The ATAC data has 22463 observations and 116490 features.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The GEX data has {adata_gex.n_obs} observations and {adata_gex.n_vars} features.\")\n",
    "print(f\"The ATAC data has {adata_atac.n_obs} observations and {adata_atac.n_vars} features.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5fb93533",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data set\n",
      "GEX data: Total cells=22463, Number features=12160\n",
      "ATAC data: Total cells=22463, Number Features=53020\n"
     ]
    }
   ],
   "source": [
    "# filter out the data\n",
    "# Convert anndata objects to dataframes and filter.  \n",
    "# Genes that show up in < 1% cells are dropped.\n",
    "# Atac seq data that shows up in < 1% cells are dropped \n",
    "\n",
    "gex_df = adata_gex.to_df()\n",
    "atac_df = adata_atac.to_df()\n",
    "\n",
    "mask = gex_df>0\n",
    "total_cells = gex_df.shape[0]\n",
    "maskdf = mask.sum(axis=0)/total_cells*100 <=1\n",
    "\n",
    "gex_feature_drop  = list(maskdf.loc[maskdf==True].index.values)\n",
    "\n",
    "mask = atac_df>0\n",
    "maskdf = mask.sum(axis=0)/total_cells <=0.01\n",
    "\n",
    "atac_feature_drop  = list(maskdf.loc[maskdf==True].index.values)\n",
    "\n",
    "gex_ = gex_df.drop(columns=gex_feature_drop)\n",
    "atac_ = atac_df.drop(columns=atac_feature_drop)\n",
    "\n",
    "print('Filtered data set')\n",
    "print('GEX data: Total cells=' + str(gex_.shape[0]) + ', Number features=' + str(gex_.shape[1]))\n",
    "print('ATAC data: Total cells=' + str(atac_.shape[0]) + ', Number Features=' + str(atac_.shape[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3e727b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AL627309.5</th>\n",
       "      <th>LINC01409</th>\n",
       "      <th>LINC01128</th>\n",
       "      <th>NOC2L</th>\n",
       "      <th>ISG15</th>\n",
       "      <th>C1orf159</th>\n",
       "      <th>SDF4</th>\n",
       "      <th>B3GALT6</th>\n",
       "      <th>UBE2J2</th>\n",
       "      <th>ACAP3</th>\n",
       "      <th>...</th>\n",
       "      <th>MT-ATP8</th>\n",
       "      <th>MT-ATP6</th>\n",
       "      <th>MT-CO3</th>\n",
       "      <th>MT-ND3</th>\n",
       "      <th>MT-ND4L</th>\n",
       "      <th>MT-ND4</th>\n",
       "      <th>MT-ND5</th>\n",
       "      <th>MT-ND6</th>\n",
       "      <th>MT-CYB</th>\n",
       "      <th>AL592183.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAGTTGTCACCCTCAC-1-s1d1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.410295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.410295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.410295</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTATGGCCATAACGGG-1-s1d1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.194758</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.194758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.168547</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGCACACAGGTTAAA-1-s1d1</th>\n",
       "      <td>0.410619</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.410619</td>\n",
       "      <td>0.821238</td>\n",
       "      <td>3.284951</td>\n",
       "      <td>0.410619</td>\n",
       "      <td>0.410619</td>\n",
       "      <td>0.821238</td>\n",
       "      <td>1.231857</td>\n",
       "      <td>0.410619</td>\n",
       "      <td>3.284951</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATTTGGTAATGGAA-1-s1d1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.879966</td>\n",
       "      <td>11.519863</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.759931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACATAGGTGTCCA-1-s1d1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.743880</td>\n",
       "      <td>13.103581</td>\n",
       "      <td>1.871940</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.743880</td>\n",
       "      <td>1.871940</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.359701</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTAAGCAACTAGGG-8-s3d6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.915288</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTCCTTCGGCTAGC-8-s3d6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTGCGTTGTTGGA-8-s3d6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.427911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCCTCCCAGCCAGTT-8-s3d6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTGAACCATCCCGCT-8-s3d6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.338799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.338799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22463 rows × 12160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         AL627309.5  LINC01409  LINC01128  NOC2L     ISG15  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1    0.000000        0.0        0.0    0.0  0.000000   \n",
       "CTATGGCCATAACGGG-1-s1d1    0.000000        0.0        0.0    0.0  0.000000   \n",
       "CCGCACACAGGTTAAA-1-s1d1    0.410619        0.0        0.0    0.0  0.000000   \n",
       "TCATTTGGTAATGGAA-1-s1d1    0.000000        0.0        0.0    0.0  0.000000   \n",
       "ACCACATAGGTGTCCA-1-s1d1    0.000000        0.0        0.0    0.0  0.000000   \n",
       "...                             ...        ...        ...    ...       ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6    0.000000        0.0        0.0    0.0  1.915288   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6    0.000000        0.0        0.0    0.0  0.000000   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6    0.000000        0.0        0.0    0.0  0.000000   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6    0.000000        0.0        0.0    0.0  0.000000   \n",
       "AGTGAACCATCCCGCT-8-s3d6    0.000000        0.0        0.0    0.0  0.000000   \n",
       "\n",
       "                         C1orf159  SDF4  B3GALT6    UBE2J2  ACAP3  ...  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1       0.0   0.0      0.0  0.000000    0.0  ...   \n",
       "CTATGGCCATAACGGG-1-s1d1       0.0   0.0      0.0  2.194758    0.0  ...   \n",
       "CCGCACACAGGTTAAA-1-s1d1       0.0   0.0      0.0  0.000000    0.0  ...   \n",
       "TCATTTGGTAATGGAA-1-s1d1       0.0   0.0      0.0  0.000000    0.0  ...   \n",
       "ACCACATAGGTGTCCA-1-s1d1       0.0   0.0      0.0  0.000000    0.0  ...   \n",
       "...                           ...   ...      ...       ...    ...  ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6       0.0   0.0      0.0  0.000000    0.0  ...   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6       0.0   0.0      0.0  0.000000    0.0  ...   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6       0.0   0.0      0.0  3.427911    0.0  ...   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6       0.0   0.0      0.0  0.000000    0.0  ...   \n",
       "AGTGAACCATCCCGCT-8-s3d6       0.0   0.0      0.0  1.338799    0.0  ...   \n",
       "\n",
       "                          MT-ATP8   MT-ATP6     MT-CO3    MT-ND3   MT-ND4L  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1  0.000000  0.000000   0.000000  4.410295  0.000000   \n",
       "CTATGGCCATAACGGG-1-s1d1  0.000000  0.000000   2.194758  0.000000  0.000000   \n",
       "CCGCACACAGGTTAAA-1-s1d1  0.410619  0.821238   3.284951  0.410619  0.410619   \n",
       "TCATTTGGTAATGGAA-1-s1d1  0.000000  2.879966  11.519863  0.000000  0.000000   \n",
       "ACCACATAGGTGTCCA-1-s1d1  0.000000  3.743880  13.103581  1.871940  0.000000   \n",
       "...                           ...       ...        ...       ...       ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6  0.000000  0.000000   0.000000  0.000000  0.000000   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6  0.000000  0.000000   0.000000  0.000000  0.000000   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6  0.000000  0.000000   0.000000  0.000000  0.000000   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6  0.000000  0.000000   0.000000  0.000000  0.000000   \n",
       "AGTGAACCATCCCGCT-8-s3d6  0.000000  1.338799   0.000000  0.000000  0.000000   \n",
       "\n",
       "                           MT-ND4    MT-ND5    MT-ND6     MT-CYB  AL592183.1  \n",
       "TAGTTGTCACCCTCAC-1-s1d1  4.410295  0.000000  0.000000   4.410295         0.0  \n",
       "CTATGGCCATAACGGG-1-s1d1  0.000000  0.000000  0.000000  13.168547         0.0  \n",
       "CCGCACACAGGTTAAA-1-s1d1  0.821238  1.231857  0.410619   3.284951         0.0  \n",
       "TCATTTGGTAATGGAA-1-s1d1  0.000000  5.759931  0.000000   0.000000         0.0  \n",
       "ACCACATAGGTGTCCA-1-s1d1  3.743880  1.871940  0.000000   9.359701         0.0  \n",
       "...                           ...       ...       ...        ...         ...  \n",
       "TAGTAAGCAACTAGGG-8-s3d6  0.000000  0.000000  0.000000   0.000000         0.0  \n",
       "TGGTCCTTCGGCTAGC-8-s3d6  0.000000  0.000000  0.000000   0.000000         0.0  \n",
       "CGCTTGCGTTGTTGGA-8-s3d6  0.000000  0.000000  0.000000   0.000000         0.0  \n",
       "ACCCTCCCAGCCAGTT-8-s3d6  0.000000  0.000000  0.000000   0.000000         0.0  \n",
       "AGTGAACCATCCCGCT-8-s3d6  0.000000  0.000000  0.000000   0.000000         0.0  \n",
       "\n",
       "[22463 rows x 12160 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gex_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "51d7f22f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1-181117-181803</th>\n",
       "      <th>chr1-629497-630394</th>\n",
       "      <th>chr1-633515-634474</th>\n",
       "      <th>chr1-778276-779191</th>\n",
       "      <th>chr1-816868-817761</th>\n",
       "      <th>chr1-827067-827948</th>\n",
       "      <th>chr1-842497-843414</th>\n",
       "      <th>chr1-869472-870377</th>\n",
       "      <th>chr1-904343-905196</th>\n",
       "      <th>chr1-906441-907357</th>\n",
       "      <th>...</th>\n",
       "      <th>GL000205.2-88673-89483</th>\n",
       "      <th>GL000205.2-140307-141166</th>\n",
       "      <th>GL000195.1-30407-31261</th>\n",
       "      <th>GL000195.1-32211-33062</th>\n",
       "      <th>GL000219.1-39933-40839</th>\n",
       "      <th>GL000219.1-42172-43054</th>\n",
       "      <th>GL000219.1-44703-45584</th>\n",
       "      <th>GL000219.1-45726-46450</th>\n",
       "      <th>GL000219.1-99257-100160</th>\n",
       "      <th>KI270713.1-21434-22336</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAGTTGTCACCCTCAC-1-s1d1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTATGGCCATAACGGG-1-s1d1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGCACACAGGTTAAA-1-s1d1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATTTGGTAATGGAA-1-s1d1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACATAGGTGTCCA-1-s1d1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTAAGCAACTAGGG-8-s3d6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTCCTTCGGCTAGC-8-s3d6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTGCGTTGTTGGA-8-s3d6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCCTCCCAGCCAGTT-8-s3d6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTGAACCATCCCGCT-8-s3d6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22463 rows × 53020 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         chr1-181117-181803  chr1-629497-630394  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                 0.0                 0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                 0.0                 1.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                 0.0                 0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                 0.0                 0.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                 0.0                 0.0   \n",
       "...                                     ...                 ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                 0.0                 0.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                 0.0                 1.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                 0.0                 0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                 0.0                 0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                 0.0                 0.0   \n",
       "\n",
       "                         chr1-633515-634474  chr1-778276-779191  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                 1.0                 0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                 1.0                 0.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                 1.0                 0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                 0.0                 0.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                 1.0                 0.0   \n",
       "...                                     ...                 ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                 0.0                 1.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                 1.0                 0.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                 0.0                 0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                 0.0                 0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                 0.0                 1.0   \n",
       "\n",
       "                         chr1-816868-817761  chr1-827067-827948  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                 0.0                 0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                 0.0                 1.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                 0.0                 0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                 0.0                 0.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                 0.0                 0.0   \n",
       "...                                     ...                 ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                 0.0                 0.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                 0.0                 1.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                 0.0                 0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                 0.0                 0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                 0.0                 0.0   \n",
       "\n",
       "                         chr1-842497-843414  chr1-869472-870377  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                 0.0                 0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                 0.0                 0.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                 0.0                 0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                 0.0                 1.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                 0.0                 0.0   \n",
       "...                                     ...                 ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                 0.0                 0.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                 0.0                 0.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                 0.0                 0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                 0.0                 0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                 0.0                 0.0   \n",
       "\n",
       "                         chr1-904343-905196  chr1-906441-907357  ...  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                 0.0                 0.0  ...   \n",
       "CTATGGCCATAACGGG-1-s1d1                 0.0                 0.0  ...   \n",
       "CCGCACACAGGTTAAA-1-s1d1                 0.0                 0.0  ...   \n",
       "TCATTTGGTAATGGAA-1-s1d1                 0.0                 0.0  ...   \n",
       "ACCACATAGGTGTCCA-1-s1d1                 0.0                 0.0  ...   \n",
       "...                                     ...                 ...  ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                 0.0                 0.0  ...   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                 1.0                 0.0  ...   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                 0.0                 1.0  ...   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                 0.0                 0.0  ...   \n",
       "AGTGAACCATCCCGCT-8-s3d6                 0.0                 0.0  ...   \n",
       "\n",
       "                         GL000205.2-88673-89483  GL000205.2-140307-141166  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                     0.0                       0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                     0.0                       0.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                     0.0                       0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                     0.0                       0.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                     0.0                       0.0   \n",
       "...                                         ...                       ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                     1.0                       0.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                     0.0                       0.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                     0.0                       0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                     0.0                       0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                     0.0                       0.0   \n",
       "\n",
       "                         GL000195.1-30407-31261  GL000195.1-32211-33062  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                     0.0                     0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                     0.0                     0.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                     1.0                     0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                     1.0                     0.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                     0.0                     0.0   \n",
       "...                                         ...                     ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                     0.0                     0.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                     0.0                     0.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                     0.0                     0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                     0.0                     0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                     0.0                     0.0   \n",
       "\n",
       "                         GL000219.1-39933-40839  GL000219.1-42172-43054  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                     0.0                     0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                     0.0                     0.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                     0.0                     0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                     0.0                     0.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                     0.0                     0.0   \n",
       "...                                         ...                     ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                     0.0                     0.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                     1.0                     0.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                     1.0                     0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                     0.0                     0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                     0.0                     0.0   \n",
       "\n",
       "                         GL000219.1-44703-45584  GL000219.1-45726-46450  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1                     0.0                     0.0   \n",
       "CTATGGCCATAACGGG-1-s1d1                     0.0                     0.0   \n",
       "CCGCACACAGGTTAAA-1-s1d1                     0.0                     0.0   \n",
       "TCATTTGGTAATGGAA-1-s1d1                     0.0                     0.0   \n",
       "ACCACATAGGTGTCCA-1-s1d1                     0.0                     0.0   \n",
       "...                                         ...                     ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6                     0.0                     0.0   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                     0.0                     0.0   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                     1.0                     0.0   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                     0.0                     0.0   \n",
       "AGTGAACCATCCCGCT-8-s3d6                     0.0                     0.0   \n",
       "\n",
       "                         GL000219.1-99257-100160  KI270713.1-21434-22336  \n",
       "TAGTTGTCACCCTCAC-1-s1d1                      1.0                     0.0  \n",
       "CTATGGCCATAACGGG-1-s1d1                      0.0                     0.0  \n",
       "CCGCACACAGGTTAAA-1-s1d1                      0.0                     0.0  \n",
       "TCATTTGGTAATGGAA-1-s1d1                      0.0                     0.0  \n",
       "ACCACATAGGTGTCCA-1-s1d1                      0.0                     0.0  \n",
       "...                                          ...                     ...  \n",
       "TAGTAAGCAACTAGGG-8-s3d6                      0.0                     0.0  \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                      0.0                     0.0  \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                      1.0                     0.0  \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                      0.0                     0.0  \n",
       "AGTGAACCATCCCGCT-8-s3d6                      0.0                     0.0  \n",
       "\n",
       "[22463 rows x 53020 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75164b1d",
   "metadata": {},
   "source": [
    "# TRANSFORMER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "710f6313",
   "metadata": {},
   "source": [
    "The Transformer T-5 small model will take in a custom dataset. This model relies solely on training therefore it is important that we have the proper pretraining before turning in this method for single cell sequencing analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1c55de98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the sequence and split it into chunks\n",
    "def read_seq_split(split_dir): \n",
    "    split_dir = Path(split_dir)\n",
    "    texts = []\n",
    "    labels = []\n",
    "    for label_dir in [\"\", \"\"]: # for each label\n",
    "        for text_file in (split_dir/label_dir).iterdir():\n",
    "            texts.append(text_file.read_text())\n",
    "\n",
    "    return texts, labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ddf4e4d",
   "metadata": {},
   "source": [
    "Loading up the data so the sequences can be read"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bac24423",
   "metadata": {},
   "source": [
    "Train our transformer using this the train_test_split() function. This wraps input validation and application to input data into a single call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6951fff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# wraps input validation and application to input data into a single call\n",
    "train_texts, val_texts, train_labels, val_labels = train_test_split(train_texts, train_labels, test_size=.2)\n",
    "\n",
    "train_encodings = tokenizer(train_texts, truncation=True, padding=True)\n",
    "val_encodings = tokenizer(val_texts, truncation=True, padding=True)\n",
    "test_encodings = tokenizer(test_texts, truncation=True, padding=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2ca87ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MIADataset(torch.utils.data.Dataset): # create a custom dataset for neurips mia model\n",
    "    def __init__(self, encodings, labels):\n",
    "        self.encodings = encodings\n",
    "        self.labels = labels\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}\n",
    "        item['labels'] = torch.tensor(self.labels[idx])\n",
    "        return item\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ff8d03",
   "metadata": {},
   "source": [
    "Using Pytorch trainer, we want to train our model. t-5 small and transformers all around rely solely on this training data so this is definatley the most important aspect of the code. Doing so will play a massive role in how we analyze this single cell data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec51714e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  assuming we want to use trainer in leiu of custom pytorch trainer\n",
    "# need to change training args based on raz input on the model\n",
    "training_args = TrainingArguments(\n",
    "    output_dir='./results',          # output directory\n",
    "    num_train_epochs=3,              # total number of training epochs\n",
    "    per_device_train_batch_size=16,  # batch size per device during training\n",
    "    per_device_eval_batch_size=64,   # batch size for evaluation\n",
    "    warmup_steps=500,                # number of warmup steps for learning rate scheduler\n",
    "    weight_decay=0.01,               # strength of weight decay\n",
    "    logging_dir='./logs',            # directory for storing logs\n",
    "    logging_steps=10,\n",
    ")\n",
    "\n",
    "model = AutoModelWithLMHead.from_pretrained(\"t5-small\")\n",
    "\n",
    "trainer = Trainer(\n",
    "    model=model,                         # the instantiated 🤗 Transformers model to be trained\n",
    "    args=training_args,                  # training arguments, defined above\n",
    "    train_dataset=train_dataset,         # training dataset\n",
    "    eval_dataset=val_dataset             # evaluation dataset\n",
    ")\n",
    "\n",
    "trainer.train()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e509739a",
   "metadata": {},
   "source": [
    "# Baseline Models given by NeurIPS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63729a4",
   "metadata": {},
   "source": [
    "a few statistical metrics given to us in the NeurIPS competition that shows how are algorithm performs. <br>\n",
    "The tests include:<br>\n",
    "**rmse** - **r**oot **m**ean **s**quare **e**rror<br>\n",
    "**baseline_linear** - linear regressor test<br>\n",
    "**baseline_mean** - mean test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e93eac05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_rmse(true_test_mod2, pred_test_mod2):\n",
    "    if pred_test_mod2.var[\"feature_types\"][0] == \"GEX\":\n",
    "        return  mean_squared_error(true_test_mod2.layers[\"log_norm\"].toarray(), pred_test_mod2.X, squared=False)\n",
    "    else:\n",
    "        raise NotImplementedError(\"Only set up to calculate RMSE for GEX data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e2834dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def baseline_linear(input_train_mod1, input_train_mod2, input_test_mod1):\n",
    "    '''Baseline method training a linear regressor on the input data'''\n",
    "    input_mod1 = ad.concat(\n",
    "        {\"train\": input_train_mod1, \"test\": input_test_mod1},\n",
    "        axis=0,\n",
    "        join=\"outer\",\n",
    "        label=\"group\",\n",
    "        fill_value=0,\n",
    "        index_unique=\"-\", \n",
    "    )\n",
    "    \n",
    "    # Binarize ATAC \n",
    "    if input_train_mod1.var[\"feature_types\"][0] == \"ATAC\":\n",
    "        input_mod1.X[input_mod1.X > 1] = 1\n",
    "    elif input_train_mod2.var[\"feature_types\"][0] == \"ATAC\":\n",
    "        input_train_mod2.X[input_mod1.X > 1] = 1\n",
    "    \n",
    "    # Do PCA on the input data\n",
    "    logging.info('Performing dimensionality reduction on modality 1 values...')\n",
    "    embedder_mod1 = TruncatedSVD(n_components=50)\n",
    "    mod1_pca = embedder_mod1.fit_transform(input_mod1.X)\n",
    "    \n",
    "    logging.info('Performing dimensionality reduction on modality 2 values...')\n",
    "    embedder_mod2 = TruncatedSVD(n_components=50)\n",
    "    mod2_pca = embedder_mod2.fit_transform(input_train_mod2.layers[\"log_norm\"])\n",
    "    \n",
    "    # split dimred mod 1 back up for training\n",
    "    X_train = mod1_pca[input_mod1.obs['group'] == 'train']\n",
    "    X_test = mod1_pca[input_mod1.obs['group'] == 'test']\n",
    "    y_train = mod2_pca\n",
    "    \n",
    "    assert len(X_train) + len(X_test) == len(mod1_pca)\n",
    "    \n",
    "    logging.info('Running Linear regression...')\n",
    "    \n",
    "    reg = LinearRegression()\n",
    "    \n",
    "    # Train the model on the PCA reduced modality 1 and 2 data\n",
    "    reg.fit(X_train, y_train)\n",
    "    y_pred = reg.predict(X_test)\n",
    "    \n",
    "    # Project the predictions back to the modality 2 feature space\n",
    "    y_pred = y_pred @ embedder_mod2.components_\n",
    "    \n",
    "    pred_test_mod2 = ad.AnnData(\n",
    "        X = y_pred,\n",
    "        obs = input_test_mod1.obs,\n",
    "        var = input_train_mod2.var,\n",
    "    \n",
    "    )\n",
    "    \n",
    "    # Add the name of the method to the result\n",
    "    pred_test_mod2.uns[\"method\"] = \"linear\"\n",
    "    \n",
    "    return pred_test_mod2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "32924350",
   "metadata": {},
   "outputs": [],
   "source": [
    "def baseline_mean(input_train_mod1, input_train_mod2, input_test_mod1):\n",
    "    '''Dummy method that predicts mean(input_train_mod2) for all cells'''\n",
    "    logging.info('Calculate mean of the training data modality 2...')\n",
    "    y_pred = np.repeat(input_train_mod2.layers[\"log_norm\"].mean(axis=0).reshape(-1,1).T, input_test_mod1.shape[0], axis=0)\n",
    "    \n",
    "    # Prepare the ouput data object\n",
    "    pred_test_mod2 = ad.AnnData(\n",
    "        X=y_pred,\n",
    "        obs=input_test_mod1.obs,\n",
    "        var=input_train_mod2.var,\n",
    "    )\n",
    "    \n",
    "    pred_test_mod2.uns[\"method\"] = \"mean\"\n",
    "\n",
    "    return pred_test_mod2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e41aa69d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "d61e67d4406f83661a218a7594034be74564666d0640d3900a3e99845865d0f0"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
